app/src/main/java/com/github/sdp/ratemyepfl/database/GradeInfoRepositoryImpl.kt

```diff
@@ -0,0 +1,111 @@
+package com.github.sdp.ratemyepfl.database
+
+import com.github.sdp.ratemyepfl.model.GradeInfo
+import com.github.sdp.ratemyepfl.model.ReviewInfo
+import com.github.sdp.ratemyepfl.model.ReviewInfo.Companion.DEFAULT_REVIEW_INFO
+import com.github.sdp.ratemyepfl.model.review.ReviewRating
+import com.google.android.gms.tasks.Task
+import com.google.firebase.firestore.DocumentSnapshot
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Transaction
+import com.google.firebase.firestore.ktx.getField
+import com.google.gson.Gson
+import com.google.gson.reflect.TypeToken
+import kotlinx.coroutines.tasks.await
+import javax.inject.Inject
+
+class GradeInfoRepositoryImpl(val repository: RepositoryImpl<GradeInfo>) : GradeInfoRepository,
+    Repository<GradeInfo> by repository {
+
+    @Inject
+    constructor(db: FirebaseFirestore) : this(RepositoryImpl<GradeInfo>(
+        db,
+        GradeInfoRepositoryImpl.GRADE_INFO_COLLECTION_PATH
+    ) {
+        it.toGradeInfo()
+    })
+
+    companion object {
+        const val GRADE_INFO_COLLECTION_PATH = "grades_info"
+        const val ITEM_ID_FIELD = "itemId"
+        const val REVIEWS_INFO_FIELD = "reviewsData"
+        const val CURRENT_GRADE_FIELD = "currentGrade"
+        const val NUM_REVIEWS_FIELD = "numReviews"
+
+        fun DocumentSnapshot.toGradeInfo(): GradeInfo? {
+            val type = object : TypeToken<Map<String, ReviewInfo>>() {}.type
+            val reviewsData = getString(REVIEWS_INFO_FIELD)?.let {
+                Gson().fromJson<Map<String, ReviewInfo>>(it, type)
+            }
+            return try {
+                GradeInfo.Builder(
+                    getString(ITEM_ID_FIELD),
+                    reviewsData,
+                    getDouble(CURRENT_GRADE_FIELD),
+                    getField<Int>(NUM_REVIEWS_FIELD)
+                ).build()
+            } catch (e: IllegalStateException) {
+                null
+            }
+        }
+    }
+
+    /**
+     * Compute the overall grade the number of reviews from all the grade info
+     * (FOR NOW BASIC AVERAGE)
+     */
+    fun computeGrade(reviewsData: Map<String, ReviewInfo>): Double {
+        var total = 0.0
```

Comment:
I think you can compute the average in one line using the method `average()`

Reply:
It's not a list, I used a map here.

Reply:
Can I also use average() on a map ?

Reply:
Yes, you can do for instance `reviewsData.values.map { it.reviewGrade }.average()`

Reply:
Yes, you can do for instance `reviewsData.values.map { it.reviewGrade }.average()`

Reply:
For now I will keep the loop, I might need to use it if I change the Math formula. (include like ratio)